File: data-processing/pulsar/sticky-telemetry-stream/sticky-telemetry-stream-publisher/src/shared/telemetry/services/publish_telemetry_stream.py
import logging
import time
from functools import partial

import pulsar
from sticky_telemetry_stream_schema.telemetry_record import TelemetryRecord

logger = logging.getLogger(__name__)


def handle_send_result(
    publisher_id: str,
    result: pulsar.Result,
    _message_id: pulsar.MessageId,
) -> None:
    if result != pulsar.Result.Ok:
        logger.error(
            f"Publisher {publisher_id}: failed to send message: {result}",
        )


def publish_telemetry_stream(
    producer: pulsar.Producer,
    publisher_id: str,
    publish_interval_s: float,
) -> int:
    published_sample_count = 0
    try:
        sample_index = 0
        while True:
            timestamp_ns = time.time_ns()

            telemetry = TelemetryRecord(
                timestamp_ns=timestamp_ns,
                temperature_c=float(sample_index),
                humidity_pct=float(sample_index),
            )

            published_sample_count += 1

            producer.send_async(
                telemetry,
                partition_key=publisher_id,
                callback=partial(handle_send_result, publisher_id),
            )

            if published_sample_count % 10 == 0:
                logger.info(
                    f"Publisher {publisher_id}: published {published_sample_count} telemetry samples",
                )

            time.sleep(publish_interval_s)
            sample_index += 1

    except KeyboardInterrupt:
        logger.info(f"Publisher {publisher_id}: telemetry streaming cancelled")
    except Exception:
        logger.exception("Failed to publish telemetry stream")
        raise

    return published_sample_count

File: data-processing/pulsar/sticky-telemetry-stream/sticky-telemetry-stream-publisher/src/shared/telemetry/services/publish_telemetry_stream.rs
use std::time::{Duration, SystemTime, UNIX_EPOCH};

use anyhow::Result;
use pulsar::producer::SendFuture;
use pulsar::{Executor, Producer};
use tracing::info;

use crate::shared::telemetry::models::telemetry_record::TelemetryRecord;

pub async fn publish_telemetry_stream<E: Executor>(
    producer: &mut Producer<E>,
    publisher_id: &str,
    publish_interval: Duration,
) -> Result<()> {
    let mut published_sample_count: u64 = 0;
    let mut sample_index: u64 = 0;
    let mut pending_send_future_list: Vec<SendFuture> = Vec::new();

    loop {
        let timestamp_ns: i64 = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .map_err(|error| anyhow::anyhow!("System clock is before UNIX epoch: {error}"))?
            .as_nanos()
            .try_into()
            .map_err(|error| anyhow::anyhow!("Timestamp nanoseconds overflow i64: {error}"))?;

        let telemetry = TelemetryRecord {
            publisher_id: publisher_id.to_string(),
            timestamp_ns,
            temperature_c: Some(sample_index as f64),
            humidity_pct: Some(sample_index as f64),
        };

        let send_future = producer
            .send_non_blocking(telemetry)
            .await
            .map_err(|error| anyhow::anyhow!("Failed to send message: {error}"))?;

        pending_send_future_list.push(send_future);
        published_sample_count += 1;

        if published_sample_count.is_multiple_of(10) {
            for send_future in pending_send_future_list.drain(..) {
                send_future.await.map_err(|error| {
                    anyhow::anyhow!("Failed to confirm message delivery: {error}")
                })?;
            }
            info!("Publisher {publisher_id}: published {published_sample_count} telemetry samples");
        }

        sample_index += 1;
        tokio::time::sleep(publish_interval).await;
    }
}
